src/main/scala/is/hail/io/bgen/BgenRecord.scala

```diff
@@ -0,0 +1,176 @@
+                if (dsum == 32768)
+                  Array(d0, d1, d2)
+                else
+                  Genotype.weightsToLinear(d0, d1, d2)
+              val gt = Genotype.unboxedGTFromLinear(px)
+              new DosageGenotype(gt, px)
+            } else {
+              setWarning(dosageGreaterThanTolerance)
+              noCall
+            }
+          } else {
+            if (dsum == 0)
+              setWarning(dosageNoCall)
+            else
+              setWarning(dosageLessThanTolerance)
+
+            noCall
+          }
+        }
+      }
+    }
+  }
+}
+
+class BgenRecordV12(compressed: Boolean, nSamples: Int, tolerance: Double) extends BgenRecord {
+  var expectedDataSize: Int = _
+  var expectedNumAlleles: Int = _
+
+  def setExpectedDataSize(size: Int) {
+    this.expectedDataSize = size
+  }
+
+  def setExpectedNumAlleles(n: Int) {
+    this.expectedNumAlleles = n
+  }
+
+  override def getValue: Iterable[Genotype] = {
+    require(input != null, "called getValue before serialized value was set")
+
+    val reader = new ByteArrayReader(if (compressed) decompress(input, expectedDataSize) else input)
+
+    val nRow = reader.readInt()
+    assert(nRow == nSamples, "row nSamples is not equal to header nSamples")
+
+    val nAlleles = reader.readShort()
+    assert(nAlleles == expectedNumAlleles, s"Value for `nAlleles' in genotype probability data storage is not equal to value in variant identifying data. Expected $expectedNumAlleles but found $nAlleles.")
+
+    val minPloidy = reader.read()
+    val maxPloidy = reader.read()
+
+    if (minPloidy != 2 || maxPloidy != 2)
+      fatal(s"Hail only supports diploid genotypes. Found min ploidy equals `$minPloidy' and max ploidy equals `$maxPloidy'.")
+
+    val samplePloidy = new Array[Int](nSamples)
+    var i = 0
+    while (i < nSamples) {
+      val ploidy = reader.read()
+      assert(ploidy == 2 || ploidy < 0, s"Ploidy value must be either 2 or missing. Found $ploidy.")
+      samplePloidy(i) = ploidy
+      i += 1
+    }
+    assert(i == nSamples, s"Number of ploidy values `$i' does not equal the number of samples `$nSamples'.")
+
+    val phase = reader.read()
+    assert(phase == 0 || phase == 1, s"Value for phase must be 0 or 1. Found $phase.")
+    val isPhased = phase == 1
+
+    if (isPhased)
+      fatal("Hail does not support phased genotypes.")
+
+    val nBitsPerProb = reader.read()
+    assert(nBitsPerProb >= 1 && nBitsPerProb <= 32, s"Value for nBits must be between 1 and 32 inclusive. Found $nBitsPerProb.")
+
+    val nGenotypes = triangle(nAlleles)
+    val nExpectedBytesProbs = (nSamples * (nGenotypes - 1) * nBitsPerProb + 7) / 8
+    assert(reader.length == nExpectedBytesProbs + nSamples + 10, s"Number of uncompressed bytes `${ reader.length }' does not match the expected size `$nExpectedBytesProbs'.")
+
+    val probabilityIterator = new BgenProbabilityIterator(reader, nBitsPerProb)
+
+    val sampleProbs = Array.fill[UInt](nGenotypes)(UInt(0))
```

Comment:
Although, I think it's fine in this case because every element is filled in with the for loop.